src/basic/fileio.c

```diff
@@ -245,26 +247,31 @@ int read_full_stream(FILE *f, char **contents, size_t *size) {
                 char *t;
                 size_t k;
 
-                t = realloc(buf, n+1);
+                t = realloc(buf, n + 1);
                 if (!t)
                         return -ENOMEM;
 
                 buf = t;
                 k = fread(buf + l, 1, n - l, f);
+                if (k > 0)
+                        l += k;
 
-                if (k <= 0) {
-                        if (ferror(f))
-                                return -errno;
+                if (ferror(f))
+                        return -errno;
 
+                if (feof(f))
                         break;
-                }
 
-                l += k;
-                n *= 2;
+                /* We aren't expecting fread() to return a short read outside
+                 * of (error && eof), assert buffer is full and enlarge buffer.
+                 */
+                assert(l == n);
 
                 /* Safety check */
-                if (n > 4*1024*1024)
+                if (n >= READ_FULL_BYTES_MAX)
```

Comment:
> Presumably the eof condition isn't known because no read() has been attempted off the end

Indeed

```sh
255                     k = fread(buf + l, 1, n - l, f);
(gdb)
256                     if (k > 0)
(gdb)
257                             l += k;
(gdb)
259                     if (ferror(f))
(gdb)
262                     if (feof(f))
(gdb)
268                     assert(l == n);
(gdb)
271                     if (n >= READ_FULL_BYTES_MAX)
(gdb) p n
$1 = 4194304
(gdb) p *f
$2 = {_flags = -72539000, _IO_read_ptr = 0x7ffff7ff4000 "", _IO_read_end = 0x7ffff7ff4000 "", _IO_read_base = 0x7ffff7ff4000 "", _IO_write_base = 0x7ffff7ff4000 "",
  _IO_write_ptr = 0x7ffff7ff4000 "", _IO_write_end = 0x7ffff7ff4000 "", _IO_buf_base = 0x7ffff7ff4000 "",
  _IO_buf_end = 0x7ffff7ff5000 "tmpfile: /tmp/test-systemd_writing_tmpfile.PFAlSkcontents: abc\nabcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789\n",
  _IO_save_base = 0x0, _IO_backup_base = 0x0, _IO_save_end = 0x0, _markers = 0x0, _chain = 0x7ffff7943560 <_IO_2_1_stderr_>, _fileno = 4, _flags2 = 64, _old_offset = 0,
  _cur_column = 0, _vtable_offset = 0 '\000', _shortbuf = "", _lock = 0x55555555e1d0, _offset = -1, __pad1 = 0x0, __pad2 = 0x55555555e1e0, __pad3 = 0x0, __pad4 = 0x0,
  __pad5 = 0, _mode = 0, _unused2 = '\000' <repeats 19 times>}
```
And we always realloc(`4Mb` + 1 )